google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/HeaderTracer.java

```diff
@@ -0,0 +1,163 @@
+import io.opencensus.stats.Measure.MeasureDouble;
+import io.opencensus.stats.Measure.MeasureLong;
+import io.opencensus.stats.MeasureMap;
+import io.opencensus.stats.Stats;
+import io.opencensus.stats.StatsRecorder;
+import io.opencensus.tags.TagContextBuilder;
+import io.opencensus.tags.TagKey;
+import io.opencensus.tags.TagValue;
+import io.opencensus.tags.Tagger;
+import io.opencensus.tags.Tags;
+import java.util.Map;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+@BetaApi
+public class HeaderTracer {
+  public static final CallOptions.Key<HeaderTracer> HEADER_TRACER_CONTEXT_KEY =
+      CallOptions.Key.create("BigtableHeaderTracer");
+  public static final CallOptions.Key<String> SPAN_NAME_CONTEXT_KEY =
+      CallOptions.Key.create("BigtableSpanName");
+
+  private Tagger tagger;
+  private StatsRecorder stats;
+  private Map<TagKey, TagValue> statsAttributes;
+
+  private HeaderTracer(Builder builder) {
+    tagger = builder.getTagger();
+    stats = builder.getStats();
+    statsAttributes = builder.getStatsAttributes();
+  }
+
+  public static class Builder {
+    private Tagger tagger;
+    private StatsRecorder stats;
+    private Map<TagKey, TagValue> statsAttributes;
+
+    private Builder() {
+      tagger = Tags.getTagger();
+      stats = Stats.getStatsRecorder();
+      statsAttributes = ImmutableMap.of();
+    }
+
+    private Builder(HeaderTracer headerTracer) {
+      tagger = headerTracer.tagger;
+      stats = headerTracer.stats;
+      statsAttributes = headerTracer.statsAttributes;
+    }
+
+    // <editor-fold desc="Public API">
+    public Builder setTagger(@Nonnull Tagger tagger) {
+      Preconditions.checkNotNull(tagger);
+      this.tagger = tagger;
+      return this;
+    }
+
+    public Builder setStats(@Nonnull StatsRecorder stats) {
+      Preconditions.checkNotNull(stats);
+      this.stats = stats;
+      return this;
+    }
+
+    public Builder setStatsAttributes(@Nonnull Map<TagKey, TagValue> statsAttributes) {
+      Preconditions.checkNotNull(statsAttributes);
+      this.statsAttributes = statsAttributes;
+      return this;
+    }
+
+    public Tagger getTagger() {
+      return tagger;
+    }
+
+    public StatsRecorder getStats() {
+      return stats;
+    }
+
+    public Map<TagKey, TagValue> getStatsAttributes() {
+      return statsAttributes;
+    }
+
+    public HeaderTracer build() {
+      Preconditions.checkNotNull(stats, "StatsRecorder must be set");
+      Preconditions.checkNotNull(tagger, "Tagger must be set");
+      Preconditions.checkNotNull(statsAttributes, "Stats attributes must be set");
+      return new HeaderTracer(this);
+    }
+    // </editor-fold>
+  }
+
+  public Tagger getTagger() {
+    return tagger;
+  }
+
+  public StatsRecorder getStats() {
+    return stats;
+  }
+
+  public Map<TagKey, TagValue> getStatsAttributes() {
+    return statsAttributes;
+  }
+
+  public void record(MeasureLong measure, long value, @Nullable String span) {
```

Comment:
I would keep this as a private method and instead expose concrete helpers that describe what you are recording: ie recordGfeLatency(String method, Duration duration)
recordMissingGfeHeader(String method)